### ISSUE TITLE
TypeError: can't cast Java::JavaSql::Timestamp

### ISSUE THREAD

**Issue by jlahtinen** [rsim/oracle-enhanced#1149]
It seems like validates_uniqueness_of to timestamp column does not work in tag 1.7.9. master branch does not have this problem.

I was able to reproduce this with following spec https://gist.github.com/jlahtinen/ea09d57def5165c7e7bf84ab85974c87

Quickly thinking it seems that in lib/active_record/connection_adapters/oracle_enhanced/jdbc_quoting.rb this can be fixed by not casting Java::JavaSql::Timestamp and just return value as it is. Fixing this like that may break something else, not sure at all.

Should I try to make a pull request about this issue applying gist and fix or only gist? If I make a pull request to what branch I should make it?

**Comment by yahonda** [comment]
Thanks for the investigation. I am also able to reproduce it.

### Environment:
* Ruby and Java version
```
$ ruby -v
jruby 9.1.7.0 (2.3.1) 2017-01-11 68056ae OpenJDK 64-Bit Server VM 25.121-b14 on 1.8.0_121-b14 +jit [linux-x86_64]
$ java -version
openjdk version "1.8.0_121"
OpenJDK Runtime Environment (build 1.8.0_121-b14)
OpenJDK 64-Bit Server VM (build 25.121-b14, mixed mode)
```
* Oracle Database 12c 12.1.0.2

* Testcase
```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", "~> 5.0.0"
  gem "activerecord", "~> 5.0.0"
  gem "activerecord-oracle_enhanced-adapter", "~> 1.7.0"
  gem "minitest"
end

require "active_record"
require "minitest/autorun"
require "logger"
require "active_record/connection_adapters/oracle_enhanced_adapter"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# Set Oracle enhanced adapter specific connection parameters
DATABASE_NAME = ENV["DATABASE_NAME"] || "orcl"
DATABASE_HOST = ENV["DATABASE_HOST"]
DATABASE_PORT = ENV["DATABASE_PORT"]
DATABASE_USER = ENV["DATABASE_USER"] || "oracle_enhanced"
DATABASE_PASSWORD = ENV["DATABASE_PASSWORD"] || "oracle_enhanced"
DATABASE_SYS_PASSWORD = ENV["DATABASE_SYS_PASSWORD"] || "admin"

CONNECTION_PARAMS = {
  adapter: "oracle_enhanced",
  database: DATABASE_NAME,
  host: DATABASE_HOST,
  port: DATABASE_PORT,
  username: DATABASE_USER,
  password: DATABASE_PASSWORD
}

ActiveRecord::Base.establish_connection(CONNECTION_PARAMS)

ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :test_logs, force: true do |t|
    t.timestamp :send_time
  end
end

class TestLog < ActiveRecord::Base
  validates_uniqueness_of :send_time
end

class BugTest < Minitest::Test
  def test_timestamp
    TestLog.create! send_time: Time.now + 1.seconds
    TestLog.create! send_time: Time.now + 2.seconds
    assert_equal 2, TestLog.count
  end
end
```

```ruby
$ ruby rep2_jruby_50_github.rb
Fetching gem metadata from https://rubygems.org/..........
Fetching version metadata from https://rubygems.org/..
Fetching dependency metadata from https://rubygems.org/.
Resolving dependencies........
Using rake 12.0.0
Using concurrent-ruby 1.0.4 (java)
Using i18n 0.8.0
Using minitest 5.10.1
Using thread_safe 0.3.5 (java)
Using builder 3.2.3
Using erubis 2.7.0
Using nokogiri 1.7.0.1 (java)
Using rack 2.0.1
Using nio4r 1.2.1 (java)
Using websocket-extensions 0.1.2
Using mime-types-data 3.2016.0521
Using arel 7.1.4
Using ruby-plsql 0.6.0
Using bundler 1.14.3
Using method_source 0.8.2
Using thor 0.19.4
Using tzinfo 1.2.2
Using loofah 2.0.3
Using rack-test 0.6.3
Using sprockets 3.7.1
Using websocket-driver 0.6.5 (java)
Using mime-types 3.1
Using activesupport 5.0.1
Using rails-html-sanitizer 1.0.3
Using mail 2.6.4
Using rails-dom-testing 2.0.2
Using globalid 0.3.7
Using activemodel 5.0.1
Using actionview 5.0.1
Using activejob 5.0.1
Using activerecord 5.0.1
Using actionpack 5.0.1
Using activerecord-oracle_enhanced-adapter 1.7.9
Using actioncable 5.0.1
Using actionmailer 5.0.1
Using railties 5.0.1
Using sprockets-rails 3.2.0
Using rails 5.0.1
-- create_table(:test_logs, {:force=>true})
D, [2017-02-02T16:38:18.004000 #1689] DEBUG -- :    (29.0ms)  DROP TABLE "TEST_LOGS"
D, [2017-02-02T16:38:18.018000 #1689] DEBUG -- :    (8.0ms)  DROP SEQUENCE "TEST_LOGS_SEQ"
D, [2017-02-02T16:38:18.045000 #1689] DEBUG -- :    (20.0ms)  CREATE TABLE "TEST_LOGS" ("ID" NUMBER(38) NOT NULL PRIMARY KEY, "SEND_TIME" TIMESTAMP)
D, [2017-02-02T16:38:18.053000 #1689] DEBUG -- :    (5.0ms)  CREATE SEQUENCE "TEST_LOGS_SEQ" START WITH 10000
   -> 1.1524s
D, [2017-02-02T16:38:18.359000 #1689] DEBUG -- :    (42.0ms)  SELECT cols.column_name AS name, cols.data_type AS sql_type, cols.data_default, cols.nullable, cols.virtual_column, cols.hidden_column, cols.data_type_owner AS sql_type_owner, DECODE(cols.data_type, 'NUMBER', data_precision, 'FLOAT', data_precision, 'VARCHAR2', DECODE(char_used, 'C', char_length, data_length), 'RAW', DECODE(char_used, 'C', char_length, data_length), 'CHAR', DECODE(char_used, 'C', char_length, data_length), NULL) AS limit, DECODE(data_type, 'NUMBER', data_scale, NULL) AS scale, comments.comments as column_comment FROM all_tab_cols cols, all_col_comments comments WHERE cols.owner = 'ORACLE_ENHANCED' AND cols.table_name = 'AR_INTERNAL_METADATA' AND cols.hidden_column = 'NO' AND cols.owner = comments.owner AND cols.table_name = comments.table_name AND cols.column_name = comments.column_name ORDER BY cols.column_id
D, [2017-02-02T16:38:18.461000 #1689] DEBUG -- :   ActiveRecord::InternalMetadata Load (43.0ms)  SELECT  "AR_INTERNAL_METADATA".* FROM "AR_INTERNAL_METADATA" WHERE "AR_INTERNAL_METADATA"."KEY" = :a1 FETCH FIRST :a2 ROWS ONLY  [["key", :environment], ["LIMIT", 1]]
Run options: --seed 1333

# Running:

D, [2017-02-02T16:38:18.762000 #1689] DEBUG -- :    (13.0ms)  SELECT cols.column_name AS name, cols.data_type AS sql_type, cols.data_default, cols.nullable, cols.virtual_column, cols.hidden_column, cols.data_type_owner AS sql_type_owner, DECODE(cols.data_type, 'NUMBER', data_precision, 'FLOAT', data_precision, 'VARCHAR2', DECODE(char_used, 'C', char_length, data_length), 'RAW', DECODE(char_used, 'C', char_length, data_length), 'CHAR', DECODE(char_used, 'C', char_length, data_length), NULL) AS limit, DECODE(data_type, 'NUMBER', data_scale, NULL) AS scale, comments.comments as column_comment FROM all_tab_cols cols, all_col_comments comments WHERE cols.owner = 'ORACLE_ENHANCED' AND cols.table_name = 'TEST_LOGS' AND cols.hidden_column = 'NO' AND cols.owner = comments.owner AND cols.table_name = comments.table_name AND cols.column_name = comments.column_name ORDER BY cols.column_id
D, [2017-02-02T16:38:18.889000 #1689] DEBUG -- :   Sequence (5.0ms)  select us.sequence_name from all_sequences us where us.sequence_owner = 'ORACLE_ENHANCED' and us.sequence_name = 'TEST_LOGS_SEQ'
D, [2017-02-02T16:38:18.902000 #1689] DEBUG -- :   Primary Key (7.0ms)  SELECT cc.column_name FROM all_constraints c, all_cons_columns cc WHERE c.owner = 'ORACLE_ENHANCED' AND c.table_name = 'TEST_LOGS' AND c.constraint_type = 'P' AND cc.owner = c.owner AND cc.constraint_name = c.constraint_name
E

Finished in 0.356706s, 2.8034 runs/s, 0.0000 assertions/s.

  1) Error:
BugTest#test_timestamp:
TypeError: can't cast Java::JavaSql::Timestamp
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/quoting.rb:40:in `type_cast'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-oracle_enhanced-adapter-1.7.9/lib/active_record/connection_adapters/oracle_enhanced/quoting.rb:146:in `type_cast'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-oracle_enhanced-adapter-1.7.9/lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:20:in `block in exec_query'
    org/jruby/RubyArray.java:2485:in `map'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-oracle_enhanced-adapter-1.7.9/lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:20:in `exec_query'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-oracle_enhanced-adapter-1.7.9/lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:82:in `select_rows'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:54:in `select_value'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/relation/finder_methods.rb:335:in `exists?'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/validations/uniqueness.rb:29:in `validate_each'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activemodel-5.0.1/lib/active_model/validator.rb:151:in `block in validate'
    org/jruby/RubyArray.java:1733:in `each'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activemodel-5.0.1/lib/active_model/validator.rb:148:in `validate'
    org/jruby/RubyKernel.java:1810:in `public_send'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:405:in `block in make_lambda'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:169:in `block in halting'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:547:in `block in default_terminator'
    org/jruby/RubyKernel.java:1120:in `catch'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:546:in `block in default_terminator'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:170:in `block in halting'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:454:in `block in call'
    org/jruby/RubyArray.java:1733:in `each'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:454:in `call'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:101:in `__run_callbacks__'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:750:in `_run_validate_callbacks'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activemodel-5.0.1/lib/active_model/validations.rb:408:in `run_validations!'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activemodel-5.0.1/lib/active_model/validations/callbacks.rb:113:in `block in run_validations!'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:97:in `__run_callbacks__'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activesupport-5.0.1/lib/active_support/callbacks.rb:750:in `_run_validation_callbacks'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activemodel-5.0.1/lib/active_model/validations/callbacks.rb:113:in `run_validations!'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activemodel-5.0.1/lib/active_model/validations.rb:338:in `valid?'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/validations.rb:65:in `valid?'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/validations.rb:82:in `perform_validations'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/validations.rb:50:in `save!'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/attribute_methods/dirty.rb:30:in `save!'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/transactions.rb:324:in `block in save!'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/transactions.rb:395:in `block in with_transaction_returning_status'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `block in transaction'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/transaction.rb:189:in `within_new_transaction'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:232:in `transaction'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/transactions.rb:211:in `transaction'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/transactions.rb:392:in `with_transaction_returning_status'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/transactions.rb:324:in `save!'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/suppressor.rb:45:in `save!'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/persistence.rb:51:in `create!'
    rep2_jruby_50_github.rb:57:in `test_timestamp'

1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
$
```

**Comment by yahonda** [comment]
Instead of `Java::JavaSql::Timestamp`, `Time` is expected.  Then rather adding `Java::JavaSql::Timestamp` to `jdbc_quoting.rb`, I prefer to diagnose why `Java::JavaSql::Timestamp` class value arrives here first.

**Comment by yahonda** [comment]
Looks like this value is casted twice. 2nd-time cast is triggered by `validates_uniqueness_of` which finally calls `exec_query`, it performs another type_cast.

```
  1) Error:
BugTest#test_timestamp:
TypeError: can't cast Java::JavaSql::Timestamp
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/quoting.rb:40:in `type_cast'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-oracle_enhanced-adapter-1.7.9/lib/active_record/connection_adapters/oracle_enhanced/quoting.rb:146:in `type_cast'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-oracle_enhanced-adapter-1.7.9/lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:20:in `block in exec_query'
    org/jruby/RubyArray.java:2485:in `map'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-oracle_enhanced-adapter-1.7.9/lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:20:in `exec_query'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-oracle_enhanced-adapter-1.7.9/lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:82:in `select_rows'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/connection_adapters/abstract/database_statements.rb:54:in `select_value'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/relation/finder_methods.rb:335:in `exists?'
    /home/yahonda/.rbenv/versions/jruby-9.1.7.0/lib/ruby/gems/shared/gems/activerecord-5.0.1/lib/active_record/validations/uniqueness.rb:29:in `validate_each'
```

Then just returning value when `Java::JavaSql::Timestamp` should be fine for release17 branch, assuming this value is already casted. For both master and release17 branch I'll add the same specs for both.

**Comment by yahonda** [comment]
#1132 removes Oracle enhanced own `select_rows` method which used to call `exec_query`. Now master branch uses Abstract adapter's `select_rows` which does not call methods which executes another type_cast.

**Comment by yahonda** [comment]
Closed by #1153 . This fix is included to Oracle enhanced adapter 1.7.10. 

Thanks for using Oracle enhanced adapter and reporting with the reproduce test case, which helps a lot.
